### ISSUE TITLE
Feature Question: Record Subtyping

### ISSUE THREAD

**Issue by srush** [google-research/dex-lang#308]
We have been playing around with using Dex record types for generic name-based contraction. The lens / zipper example is clear about how make this work. 

It does immediately make me wonder though if you can do broadcasting / subtyping on records. In particular given two record index types it would be nice to write a function like this: 

```haskell
def promote (tab: {...b} => out) : {a & ...b} => out =                                    
    for i : {a & ...b}.                                                                 
        tab.(down i) 
```
I thought a bit about how to do it, but was completely stumped. I realize that you can do it by providing the lens explicitly, but that is less satisfying. 

```haskell
def ndim (name: Iso a (b & c))                                                       
         (tensor: c => out) : (a => out) =                                           
    push name for i: b. tensor 
```

I see that this does exist in http://hackage.haskell.org/package/generic-lens-2.0.0.0/docs/Data-Generics-Product-Subtype.html , but perhaps it requires some trickery. 

Seems related to https://github.com/google-research/dex-lang/issues/258

**Comment by danieldjohnson** [comment]
Record subtyping would definitely be cool to have! Indeed, it probably would depend on something like #258 or a separate special-case in the compiler.

In this case, though, I don't think you need record subtyping if you instead use (variadic) unpacking:
```
def promote (rest: Fields)?-> (tab: {&...rest} => out) : ({a:a & ...rest} => out) =                                    
    for {a=_, ...restval}.
        -- restval is a record of type {&...rest}
        tab.restval
```

Admittedly the syntax for this is fairly non-obvious at the moment!

**Comment by srush** [comment]
Thanks! This is really neat. However, this only works over a record with key named "a:"? Ideally it would use whatever name is used in the output type. I'm guess that is not currently possible.

(Also hi! Neat to talk to you in two different contexts this week. )

**Comment by danieldjohnson** [comment]
Yeah, my suggestion requires you to know statically which keys you are broadcasting over. The general subtype-based approach might not have that limitation, but I'm not sure how type inference would work in that case (and I think it's something Dex's current typechecker can't handle).

(And hi 👋 !)

**Comment by srush** [comment]
Gotcha. Yeah, I figured this was hard, but the lens stuff was surprising enough to me that I thought there might be another trick I was missing. 

I think this sort of structural subtyping is rare but does exist in some ML like languages, so it might be possible? It's obviously super common in how numpy folks do casual inferences. (And of course it would be really badass to have statically typed broadcasting).

**Comment by apaszke** [comment]
When you say that you want statically typed broadcasting, isn't the lens/iso-based implementation what you want? I understand that specifying the field names to broadcast along might be a good thing, but I might be missing something.

Also, note that in many cases broadcasting in Dex is unnecessary and can be replaced by a loop. Instead of enlarging all inputs to match the same size, you can iterate over the domain and reuse the same value at each iteration. Here's an example:
```
-- x : { features: Fin n }, y : { batch: Fin b, features: Fin n }
broadcast #batch x + y
-- vs.
for b. x + y.b
```

**Comment by oxinabox** [comment]
I am a big fan of the fact the Dex neither needs numpy-style implict broadcasting,
nor julia-style explict broadcasting because the loop construct is so compact and so clean

**Comment by srush** [comment]
Thanks @apaszke. I agree that what you describe is really ideal. 

However, I am not sure I understand how the type inference works. When I run your example I get an error. 

```haskell
def (++) (tab1 : a=>Float) (tab2 : a =>Float) : a => Float =                         
   for i. tab1.i + tab2.i                                                            
                                                                                                                                                                 
x1: {a: Fin 20 & b: Fin 20} => Float = for i. 10.0                                   
x2 : {b: Fin 20} => Float = for i. 10.0                                              
                                                                                     
x3 = for i. x1.i ++ x2
```

**Comment by apaszke** [comment]
@srush yeah, good point. This works when the dimensions are uncurried (i.e. `n=>m=>...=>Float`), but not when you want to modify the type of the outer dimension (e.g. add another field), because `for` _always_ adds an extra outer dimension. So I think that you might have to use the record unpacking and index out all necessary components of the tables:
```
x3 = for {b, ...rest}. x1.{b, ...rest} + x2.{b}
```

Maybe the takeaway here is that array-level functions are quite inconvenient to use in Dex and in most places you would just end up unpacking all the way to the scalar level, and using the scalar `+` instead of any vectorized version.

------

A potential alternative would be to have a dedicated indexing operator that indexes out a single dimension of a tuple:
```
def indexByName (name: Iso full (n & rest)) (tab : full=>a) (idx: n) : rest=>a =
  for i. tab.(pushAt name idx i)

x4 = for i. indexByName #a x1 i ++ x2
```
But then, of course, the type of `x4` is `Fin 20=>{b : Fin 20}=>Float`, which is not what we want 😕

**Comment by srush** [comment]
Right. Okay, well not to worry, unpacking is awesome. (I feel like records only kind of half work in haskell anyway, currying is much cleaner.)

Just for fun it is interesting to imagine an unordered  `rfor i:{a:a}. x.a: ({a & ...rest}=>Float)` that  produces `{a & ...rest}=>Float`. I can only imagine though how that would break the internals.
